### PR TITLE
feat: register masternodes on testnet given funding privkey

### LINKED ISSUE #284: register masternode in setup command on testnet
Since developers will be working on testnet to prepare platform clients, they will need to be able to set up and register masternodes locally as well. A warning as suggested in #280 should be included, and several checks to ensure the function cannot be used on mainnet.

## Expected Behavior
`mn setup testnet masternode` should offer to register the masternode as well, given a private key with sufficient balance.

## Current Behavior
it is not currently possible to register masternodes on testnet

### PR DESCRIPTION
**Pull request by strophy**
This PR moves the register function into the setup function and enables register functionality for testnet masternodes, if the user supplies a private key. Resolves #284 

## Issue being fixed or feature implemented
New developers needing direct access to masternodes but otherwise unfamiliar with Dash were not able to get started quickly using mn-bootstrap due to difficulty with registration. This functionality should simplify the process.

## What was done?
- Removed `register` command
- Added option to provide funding private key to setup command
- Enabled register step for testnet if key is provided
- Removed BLS key generation from `registerMasternodeTask` (already exists in `setup` command)
- Added progress output for masternode sync

## How Has This Been Tested?
Can't test local because 0.19-dev doesn't work for me. Tested with `testnet` config, masternode comes up and platform sync begins.

## Breaking Changes
- `register` command removed

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by strophy** [comment]
Linux test result:
```
strophy@ubuntu-s-2vcpu-2gb-intel-sfo3-01:~/mn-bootstrap$ mn setup testnet masternode -p cRgLzcdAAjxKchTS65YdGKq53M5jKEo17Y7J316N7uJeD14RGJ5f
✔ Set configuration preset
✔ Set node type
  › Selected masternode type
✔ Configure external IP address
  › 143.198.57.241 is set
✔ Set masternode operator private key
  › BLS public key: 0263eb57ee231de125112dca758cbcaa9006d5ceb99c260b6f40deb6852ea0f4983c4c4253de4029137abcaa8f437c70
    BLS private key: 6c5f4ac9fd9925de1f04cdab1748be7ce7d9149d21f42b08cc3e3150a252ada2
✔ Set up funding key
✔ Register masternode
  ✔ Start Core
  ✔ Import funding private key
  ✔ Sync Core with network
    › 100.00% complete
  ✔ Check funding address yNff5b8ZTPMWQqiVUgD5BJXftgDPWPLHFh balance
  ✔ Create a new collateral address
    › Address: yhXJR44cMAh3pj83YoUjs4MhdnK5xhQmGp
      Private key: cMfS1Kn5bEcuppDxBEv5eaGCz2B1B4Erd7mn2GQcw2Qjhro6v59H
  ✔ Create a new owner addresses
    › Address: yLgEzVG5qNcetaQdEHa4oiqcdVFhph5eoq
      Private key: cTFWd6tXJyn4YgSSEHBQrHw26LSztZAFEwdQ7PhyQEuamRTsCe3P
  ✔ Send 1000 dash from funding address to collateral address
    › Collateral transaction ID: f1379bacb19dd00d32bc5b073ee1530711eedec7e8edd3d85e9396a662dff97d
  ✔ Wait for 1 confirmation
  ✔ Broadcast masternode registration transaction
    › ProRegTx transaction ID: 946b28571cd4d593d20707c88d8c65deec77925e76723dc42b0d78be3e79bb01
  ✔ Wait for 1 confirmation
  ✔ Stop Core
✔ Initialize Tenderdash
  ✔ Generate node keys and data
✔ Set default config
strophy@ubuntu-s-2vcpu-2gb-intel-sfo3-01:~/mn-bootstrap$ mn start
✔ Start testnet node
  ✔ Check node is not started
  ✔ Start services
strophy@ubuntu-s-2vcpu-2gb-intel-sfo3-01:~/mn-bootstrap$ mn status:services
╔═════════════════════════════════╤══════════════╤════════════════════════════════════╤═════════╗
║ Service                         │ Container ID │ Image                              │ Status  ║
╟─────────────────────────────────┼──────────────┼────────────────────────────────────┼─────────╢
║ Core                            │ 66f4bc6b98d5 │ dashpay/dashd:0.17.0.0-rc3-hotfix1 │ running ║
║ Sentinel                        │ cf130bce713f │ dashpay/sentinel:1.5.0             │ running ║
║ Drive MongoDB                   │ 0b48628ecf31 │ mongo:4.2                          │ running ║
║ Drive ABCI                      │ b3162f28abe9 │ dashpay/drive:0.18-dev             │ running ║
║ Drive Tenderdash                │ 8ec3264c3934 │ dashpay/tenderdash:0.34.3          │ running ║
║ DAPI Insight                    │ 2bcb21976ff5 │ dashpay/insight-api:3.1.1          │ running ║
║ DAPI API                        │ e340ce32833d │ dashpay/dapi:0.18-dev              │ running ║
║ DAPI Transactions Filter Stream │ c13099d4a383 │ dashpay/dapi:0.18-dev              │ running ║
║ DAPI Envoy                      │ 97ce2419b8e6 │ envoyproxy/envoy:v1.16-latest      │ running ║
║ DAPI Nginx                      │ 9c9aea2677e9 │ nginx:latest                       │ running ║
╚═════════════════════════════════╧══════════════╧════════════════════════════════════╧═════════╝

strophy@ubuntu-s-2vcpu-2gb-intel-sfo3-01:~/mn-bootstrap$ mn status
╔═══════════════════╤═══════════════════╗
║ Network           │ test              ║
║ Core Version      │ 0.17.0.0          ║
║ Core Status       │ running           ║
║ Masternode Status │ Ready             ║
║ Platform Version  │                   ║
║ Platform Status   │ syncing Infinity% ║
╚═══════════════════╧═══════════════════╝
```

**Comment by strophy** [comment]
Windows test result - also requires merging #291 
image

File: src/core/waitForCoreSync.js [hunk 1]
```diff
@@ -5,16 +5,26 @@ const wait = require('../util/wait');
  *
  * @typedef {waitForCoreSync}
  * @param {CoreService} coreService
+ * @param {task} task
  * @return {Promise<void>}
  */
-async function waitForCoreSync(coreService) {
+async function waitForCoreSync(coreService, task) {
```
Comment (shumkov) [inline src/core/waitForCoreSync.js:11]: It’s better to pass a callback and call it with data so we can decouple the listr and this function.

**Review by shumkov — CHANGES_REQUESTED** [review]

File: src/core/waitForCoreSync.js (hunk 1 above)
Comment (strophy) [reply src/core/waitForCoreSync.js]: OK, I think I did it right. Testing this was strange - previously the `Sync Core with network` step would wait for core to sync, but passing the callback function caused `waitForCoreSync()` to return immediately, even though it wasn't complete yet. I had to add `await` to the function call. Why?

File: src/listr/tasks/registerMasternodeTaskFactory.js [hunk 2]
```diff
@@ -67,11 +60,19 @@ function registerMasternodeTaskFactory(
...
       {
         title: 'Sync Core with network',
         enabled: () => config.get('network') !== NETWORK_LOCAL,
-        task: async (ctx) => waitForCoreSync(ctx.coreService),
+        task: async (ctx, task) => {
+          function updateSyncStatus(verificationProgress) {
+            // eslint-disable-next-line no-param-reassign
+            task.output = `${(verificationProgress * 100).toFixed(2)}% complete`;
+          }
+          await waitForCoreSync(ctx.coreService, updateSyncStatus);
```
Comment (shumkov) [inline src/listr/tasks/registerMasternodeTaskFactory.js:73]: There is no need to pass a named function. You can pass just a closure (array function).

File: src/core/waitForCoreSync.js [hunk 3]
```diff
@@ -5,16 +5,24 @@ const wait = require('../util/wait');
  *
  * @typedef {waitForCoreSync}
  * @param {CoreService} coreService
+ * @param {updateSyncStatus} updateSyncStatus
  * @return {Promise<void>}
  */
-async function waitForCoreSync(coreService) {
+async function waitForCoreSync(coreService, updateSyncStatus) {
```
Comment (shumkov) [inline src/core/waitForCoreSync.js:11]: Usually, such things called callbacks. In this case, it's `progressCallback`

File: src/listr/tasks/registerMasternodeTaskFactory.js [hunk 4]
```diff
@@ -135,14 +124,14 @@ function registerMasternodeTaskFactory(
         options: { persistentOutput: true },
       },
       {
-        title: 'Wait for 15 confirmations',
+        title: 'Wait for 1 confirmation',
```
Comment (shumkov) [inline src/listr/tasks/registerMasternodeTaskFactory.js:127]: Are you sure that 1 confirmation is enough? We had 15 for some reason. Maybe it's a regtest specific thing?

File: src/listr/tasks/registerMasternodeTaskFactory.js (hunk 2 above)
Comment (strophy) [reply src/listr/tasks/registerMasternodeTaskFactory.js]: OK, thanks for the above explanations. Now I was finally able to match the pattern used in the other functions. But I still can't understand _why_ it works...

File: src/listr/tasks/registerMasternodeTaskFactory.js (hunk 4 above)
Comment (shumkov) [reply src/listr/tasks/registerMasternodeTaskFactory.js:133]: I got it. Thank you.

File: src/core/waitForCoreSync.js [hunk 5]
```diff
@@ -5,16 +5,24 @@ const wait = require('../util/wait');
  *
  * @typedef {waitForCoreSync}
  * @param {CoreService} coreService
+ * @param {updateSyncStatus} updateSyncStatus
```
Comment (shumkov) [inline src/core/waitForCoreSync.js:8]: Please update doc  block

Comment (strophy) [reply src/core/waitForCoreSync.js:8]: OK, updated. Is JSDoc part of TypeScript? Or what is the purpose of this section?

**Review by shumkov — APPROVED** [review]
LGTM! 👍

**Review by shumkov — CHANGES_REQUESTED** [review]
Oh.. We forgot to revisit "Register masternode" section in README.

**Review by shumkov — APPROVED** [review]
👍

**Pull request merged**